endpoint-common/src/command/source.rs

```diff
@@ -0,0 +1,88 @@
+use crate::{
+    command::{Command, CommandDispatcher},
+    kafka::KafkaConfig,
+};
+use async_trait::async_trait;
+use drogue_cloud_event_common::stream::{
+    AutoAck, EventStream, EventStreamConfig, EventStreamError,
+};
+use drogue_cloud_service_api::health::{HealthCheckError, HealthChecked};
+use drogue_cloud_service_common::defaults;
+use futures::StreamExt;
+use serde::Deserialize;
+use std::{
+    convert::TryFrom,
+    sync::atomic::{AtomicBool, Ordering},
+};
+use tokio::task::JoinHandle;
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct KafkaCommandSourceConfig {
+    #[serde(default, flatten)]
+    pub kafka: KafkaConfig,
+    #[serde(default = "defaults::kafka_command_topic")]
+    pub topic: String,
+    pub consumer_group: String,
+}
+
+pub struct KafkaCommandSource {
+    handle: JoinHandle<()>,
+    alive: AtomicBool,
+}
+
+impl KafkaCommandSource {
+    pub fn new<D>(dispatcher: D, config: KafkaCommandSourceConfig) -> Result<Self, EventStreamError>
+    where
+        D: CommandDispatcher + Send + Sync + 'static,
+    {
+        let mut source = EventStream::<AutoAck>::new(EventStreamConfig {
+            bootstrap_servers: config.kafka.bootstrap_servers,
+            properties: config.kafka.custom,
+            topic: config.topic,
+            consumer_group: Some(config.consumer_group),
+        })?;
+
+        let handle = tokio::spawn(async move {
+            while let Some(event) = source.next().await {
+                match event {
+                    Ok(event) => match Command::try_from(event) {
+                        Ok(command) => {
+                            if let Err(err) = dispatcher.send(command).await {
+                                log::info!("Failed to dispatch command: {}", err);
+                            }
+                        }
+                        Err(_) => {
+                            log::info!("Failed to convert event to command");
+                        }
+                    },
+                    Err(err) => {
+                        log::info!("Failed to read next event: {}", err);
+                    }
```

Comment:
Should these bubble up? Or you wanna deal with that later?

Reply:
Well, there is nothing "above" it. As it is running as tokio task, no one else would care.

My idea was (and maybe I should write that in the source code) … if there is a command which cannot be handled, that can happen. And we just discard it. And someone would need to re-try.